### PR TITLE
Doc: Add Developer Guide with internals on the testing package

### LINKED ISSUE #1210: Add internal design documentation for UITester
This issue is for adding documentation on the design of the UITester to the public sphinx doc.
The main target audience are current and future contributors who want to modify and/or maintain the testing library.

### PR DESCRIPTION
**Pull request by kitchoi**
Closes #1210 

This PR adds a new section in the documentation for current and future developers who maintain / contribute to TraitsUI, and those who simply want to understand more of the internal structures. Then it adds a section to document the design and motivation for the `traitsui.testing` package.

### CODE REVIEW

File: docs/source/traitsui_dev_guide/testing.rst [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+
+The design was rejected for two reasons:
+
+#. The control flow resulting from having to try resolving an interaction
+   or location on the default target adds complexity to the code.
+#. The resulting code is obscure because it is hard to see which UI editors
+   are using the interaction handlers registered. This makes it difficult to
+   perform impact analysis when one needs to change the code.
+
+At the end we simple register the interaction types on those simple UI editors
```
Comment (rahulporuri) [inline docs/source/traitsui_dev_guide/testing.rst:256]:
```suggestion
At the end we simply register the interaction types on those simple UI editors
```

File: docs/source/traitsui_dev_guide/testing.rst [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+--------------------
+
+When a functionality is added to support testing TraitsUI editors, the
+functionality should be used in TraitsUI 's own tests for the editor.
+e.g. The MouseClick support for ButtonEditor is tested in the tests for
+ButtonEditor, in ``traitsui.editors.tests.test_button_editor``.
+
+By dog-fooding the implementations back to TraitsUI 's own tests with the goal
+of writing toolkit independent test code, this allows us to capture
+inconsistencies among different toolkit early and to make sure the new
```
Comment (rahulporuri) [inline docs/source/traitsui_dev_guide/testing.rst:269]:
```suggestion
inconsistencies among different toolkits early and to make sure the new
```

File: docs/source/traitsui_dev_guide/testing.rst [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+functionality is indeed useful.
+
+Tests for functionality regarding the tester API should be found in the
```
Comment (rahulporuri) [inline docs/source/traitsui_dev_guide/testing.rst:272]:
```suggestion
Tests for functionality provided by the tester API can be found in the
```

**Review by rahulporuri — APPROVED** [review]
LGTM. Thanks a lot for putting this together. I have a few minor suggestions.

**Comment by rahulporuri** [comment]
Note that there's a way to batch suggestions into a single commit if you want to accept my suggestions without ruining/flooding the commit history/CI https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes

**Comment by kitchoi** [comment]
Thank you @rahulporuri 
At the end I batch committed some from the suggested changes but some of them I implemented locally to apply variations to do with punctuation / wrapping. 

These are the only two commits which are not punctuation level sort of changes:
https://github.com/enthought/traitsui/pull/1314/commits/ade2a1b99c3735a9592eed28287de772b0eab45e
https://github.com/enthought/traitsui/pull/1314/commits/6536be532bfc505589794a615b49c20af8351a6d <- address one of your comments

**Comment by rahulporuri** [comment]
Still LGTM.

**Pull request merged**